File: templates/recursion/recursion.go
package recursion

func Pow(x float64, n int) float64 {
	if n == 0 {
		return 1
	}
	if n < 0 {
		return 1 / Pow(x, -n)
	}
	half := Pow(x, n/2)
	if n%2 == 0 {
		return half * half
	}
	return half * half * x
}

func GenerateParenthesis(n int) []string {
	result := []string{}
	path := make([]byte, 0, 2*n)

	var dfs func(int, int)
	dfs = func(open, close int) {
		if len(path) == 2*n {
			result = append(result, string(path))
			return
		}
		if open < n {
			path = append(path, '(')
			dfs(open+1, close)
			path = path[:len(path)-1]
		}
		if close < open {
			path = append(path, ')')
			dfs(open, close+1)
			path = path[:len(path)-1]
		}
	}

	dfs(0, 0)
	return result
}

File: templates/recursion/recursion.py
def pow_fast(x: float, n: int) -> float:
    if n == 0:
        return 1.0
    if n < 0:
        return 1 / pow_fast(x, -n)
    half = pow_fast(x, n // 2)
    if n % 2 == 0:
        return half * half
    return half * half * x


def generate_parenthesis(n: int) -> list[str]:
    result: list[str] = []
    path: list[str] = []

    def dfs(opened: int, closed: int) -> None:
        if len(path) == 2 * n:
            result.append("".join(path))
            return
        if opened < n:
            path.append("(")
            dfs(opened + 1, closed)
            path.pop()
        if closed < opened:
            path.append(")")
            dfs(opened, closed + 1)
            path.pop()

    dfs(0, 0)
    return result
